app/scripts/background-controller.js

```diff
@@ -0,0 +1,556 @@
+        const tabId = remotePort.sender.tab.id;
+        this.openMetamaskTabsIDs[tabId] = true;
+
+        endOfStream(portStream, () => {
+          this.endStreamFullScreenCb(tabId);
+        });
+      }
+    } else if (
+      senderUrl &&
+      senderUrl.origin === this.phishingPageUrl.origin &&
+      senderUrl.pathname === this.phishingPageUrl.pathname
+    ) {
+      const portStream = new PortStream(remotePort);
+      this.controller.setupPhishingCommunication({
+        connectionStream: portStream,
+      });
+    } else {
+      if (remotePort.sender && remotePort.sender.tab && remotePort.sender.url) {
+        const tabId = remotePort.sender.tab.id;
+        const url = new URL(remotePort.sender.url);
+        const { origin } = url;
+
+        remotePort.onMessage.addListener((msg) => {
+          this.remotePortOnMessageRequestAccountsCb(msg, origin, tabId);
+        });
+      }
+      this.connectExternal(remotePort);
+    }
+  }
+
+  remotePortOnMessageRequestAccountsCb(msg, origin, tabId) {
+    if (msg.data && msg.data.method === 'eth_requestAccounts') {
+      this.requestAccountTabIds[origin] = tabId;
+    }
+  }
+
+  endStreamPopUpCb() {
+    this.popupIsOpen = false;
+    const isClientOpen = this.isClientOpenStatus();
+    this.controller.isClientOpen = isClientOpen;
+    this.onCloseEnvironmentInstances(isClientOpen, ENVIRONMENT_TYPE_POPUP);
+  }
+
+  endStreamNotificationCb() {
+    this.notificationIsOpen = false;
+    const isClientOpen = this.isClientOpenStatus();
+    this.controller.isClientOpen = isClientOpen;
+    this.onCloseEnvironmentInstances(
+      isClientOpen,
+      ENVIRONMENT_TYPE_NOTIFICATION,
+    );
+  }
+
+  endStreamFullScreenCb(tabId) {
+    delete this.openMetamaskTabsIDs[tabId];
+    const isClientOpen = this.isClientOpenStatus();
+    this.controller.isClientOpen = isClientOpen;
+    this.onCloseEnvironmentInstances(isClientOpen, ENVIRONMENT_TYPE_FULLSCREEN);
+  }
+
+  // communication with page or other extension
+  connectExternal(remotePort) {
+    const portStream = new PortStream(remotePort);
+    this.controller.setupUntrustedCommunication({
+      connectionStream: portStream,
+      sender: remotePort.sender,
+    });
+  }
+
+  /**
+   * Updates the Web Extension's "badge" number, on the little fox in the toolbar.
+   * The number reflects the current number of pending transactions or message signatures needing user approval.
+   */
+  updateBadge() {
+    let label = '';
+    const count = this.getUnapprovedTransactionCount();
+    if (count) {
+      label = String(count);
+    }
+    // browserAction has been replaced by action in MV3
+    if (isManifestV3) {
+      browser.action.setBadgeText({ text: label });
+      browser.action.setBadgeBackgroundColor({ color: '#037DD6' });
+    } else {
+      browser.browserAction.setBadgeText({ text: label });
+      browser.browserAction.setBadgeBackgroundColor({ color: '#037DD6' });
+    }
```

Comment:
non-functional refactor:
```suggestion
    const count = this.getUnapprovedTransactionCount();
    const label = String(count || '');

    // browserAction has been replaced by action in MV3
    const action = isManifestV3 
      ? browser.action
      : browser.browserAction;

    action.setBadgeText({ text: label });
    action.setBadgeBackgroundColor({ color: '#037DD6' });
```